typescript/phoenix-sdk/src/utils/market.ts

```diff
@@ -812,52 +812,239 @@ export function getMarketSwapOrderPacketWithTimeInForce({
+ * This function represents a Sell order where the caller knows the amount of quote units they want to obtain but does not know how many raw base units they need to sell in order to obtain that amount of quote units.
+ *
+ * @param uiLadder The uiLadder for the market. Note that prices in the uiLadder are in quote units per raw base unit (units of USDC per unit of SOL for the SOL/USDC market).
+ * @param takerFeeBps The taker fee in bps
+ * @param quoteUnitsOut The amount of quote units to obtain
+ *
+ */
+export function getRawBaseUnitsInFromQuoteUnitsOut({
+  uiLadder,
+  takerFeeBps,
+  quoteUnitsOut,
+}: {
+  uiLadder: UiLadder;
+  takerFeeBps: number;
+  quoteUnitsOut: number;
+}): number {
+  return getBaseAmountFromQuoteAmountBudgetAndBook({
+    sideOfBook: uiLadder.bids,
+    quoteAmountBudget: quoteUnitsOut / (1 - takerFeeBps / 10000),
+  });
+}
+
+/**
+ * Given a desired amount of raw base units to obtain, return the number of quote units that need to be spent to get that amount of raw base units.
+ * This function represents a Buy order where the caller knows the amount of raw base units they want to obtain but does not know how many quote units they need to spend in order to obtain that amount of raw base units.
+ *
+ * @param uiLadder The uiLadder for the market. Note that prices in the uiLadder are in quote units per raw base unit (units of USDC per unit of SOL for the SOL/USDC market).
+ * @param takerFeeBps The taker fee in bps
+ * @param rawBaseUnitsOut The amount of raw base units to obtain
+ *
+ */
+export function getQuoteUnitsInFromRawBaseUnitsOut({
+  uiLadder,
+  takerFeeBps,
+  rawBaseUnitsOut,
+}: {
+  uiLadder: UiLadder;
+  takerFeeBps: number;
+  rawBaseUnitsOut: number;
+}): number {
+  // Walk through the asks first and find quote units to match
+  const quoteUnitsToMatch = getQuoteAmountFromBaseAmountBudgetAndBook({
+    sideOfBook: uiLadder.asks,
+    baseAmountBudget: rawBaseUnitsOut,
+  });
+  // Amount for quote units should account for fee
+  return quoteUnitsToMatch * (1 + takerFeeBps / 10000);
+}
+
+/**
+ * Given a budget of quote units in, return the number of raw base units that can be matched.
+ *
+ * @param sideOfBook The side of the book to match against. If the order is a buy order, then the sideOfBook input is the asks array. If the order is a sell order, then the sideOfBook input is the bids array.
+ * @param quoteAmountBudget The amount of quote units to match with.
+ *
+ */
+export function getBaseAmountFromQuoteAmountBudgetAndBook({
+  sideOfBook,
+  quoteAmountBudget,
+}: {
+  sideOfBook: [number, number][];
+  quoteAmountBudget: number;
+}): number {
+  // Number returned is raw base units
+  let quoteAmountBudgetRemaining = quoteAmountBudget;
+  let baseAmount = 0;
+  for (const [priceInQuoteUnitsPerBaseUnit, sizeInRawBaseUnits] of sideOfBook) {
+    if (
+      priceInQuoteUnitsPerBaseUnit * sizeInRawBaseUnits >
```

Comment:
slightly prefer >= so we don't do another loop iteration unnecessarily in the equals case